src/coprocessor_v2/plugin_registry.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use coprocessor_plugin_api::{
+    allocator::HostAllocatorPtr, CoprocessorPlugin, PluginConstructorSignature,
+    PLUGIN_CONSTRUCTOR_NAME,
+};
+use libloading::{Error as DylibError, Library, Symbol};
+use std::collections::BTreeMap;
+use std::ffi::OsStr;
+use std::marker::PhantomPinned;
+use std::pin::Pin;
+
+#[derive(Default)]
+pub struct PluginRegistry {
+    /// Plugins that are currently loaded.
+    /// Provides a mapping from the plugin's name to the actual instance.
+    loaded_plugins: BTreeMap<String, LoadedPlugin>,
+}
+
+impl PluginRegistry {
+    /// Creates a new `PluginRegistry`.
+    pub fn new() -> Self {
+        PluginRegistry::default()
+    }
+
+    /// Finds a plugin by its name. The plugin must have been loaded before with [`load_plugin()`].
+    ///
+    /// Plugins are indexed by the name that is returned by [`CoprocessorPlugin::name()`].
+    pub fn get_plugin(&self, plugin_name: &str) -> Option<&dyn CoprocessorPlugin> {
+        self.loaded_plugins.get(plugin_name).map(|p| p.plugin())
+    }
+
+    /// Loads a [`CoprocessorPlugin`] from a `dylib`.
+    ///
+    /// After this function has successfully finished, the plugin is registered with the
+    /// [`PluginRegistry`] and can later be obtained by calling [`get_plugin()`] with the proper
+    /// name.
+    ///
+    /// Returns the name of the loaded plugin.
+    pub fn load_plugin<P: AsRef<OsStr>>(
+        &mut self,
+        filename: P,
+    ) -> Result<&'static str, DylibError> {
+        let lib = unsafe { Library::new(filename)? };
+        let plugin = unsafe { LoadedPlugin::new(lib)? };
+        let plugin_name = plugin.plugin().name();
+
+        self.loaded_plugins.insert(plugin_name.to_string(), plugin);
+        Ok(plugin_name)
+    }
+}
+
+/// A wrapper around a loaded raw coprocessor plugin library.
+struct LoadedPlugin {
+    /// Pointer to a [`CoprocessorPlugin`] in the loaded `lib`.
+    plugin: Box<dyn CoprocessorPlugin>,
+    /// Underlying library file on a fixed position on the heap.
+    lib: Pin<Box<Library>>,
```

Comment:
Why the `Library` has to be pinned? Doesn't it allocate symbols in heap? (https://github.com/nagisa/rust_libloading/blob/7fbd3a34336f7839bdd6d6aaf83a9a763c4fa1e5/src/os/unix/mod.rs#L78)

Reply:
And in fact, I'll suggest to leak the Library on loading the plugin because we can't control the reference to the library once the `Future` created by `CoprocessorPlugin` is spawned into a runtime, and then it's be certainly unsafe to unmount the library.

Reply:
Oh true that! I somehow assumed that the library has it's own memory space because I never worked with dylibs before. But makes sense, thank you for the clarification! :)